frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -2034,6 +2038,32 @@ func validateStageNames(stages []instructions.Stage, warn linter.LintWarnFunc) {
 	}
 }
 
+func validateCommandArgs(allDispatchStates *dispatchStates, shlex *shell.Lex, warn linter.LintWarnFunc) {
+	nonEnvArgs := map[string]struct{}{}
+	for _, d := range allDispatchStates.states {
```

Comment:
@colinhemmings I think this should do the validation during dispatch when the base images have been resolved and it is known what variables are available. If there is a need to check all Dockerfile stages with (some config of) `--print=lint` then that can be some separate call path that "builds" all the targets in parallel.